reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -56,87 +64,73 @@ public int getPrefetch() {
 
         volatile int wip;
         @SuppressWarnings("rawtypes")
-        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstInner> WIP =
-                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstInner.class, "wip");
+        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstMain> WIP =
+                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstMain.class, "wip");
 
         volatile int state;
         @SuppressWarnings("rawtypes")
-        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstInner> STATE =
-                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstInner.class, "state");
+        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstMain> STATE =
+                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstMain.class, "state");
 
         @SuppressWarnings("unchecked")
-        AbstractSwitchOnFirstInner(
+        AbstractSwitchOnFirstMain(
                 CoreSubscriber<? super R> outer,
-                BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer) {
+                BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer,
+                boolean cancelSourceOnComplete) {
             this.outer = outer instanceof Fuseable.ConditionalSubscriber
-                ? new SwitchOnFirstConditionalInnerSubscriber<>(this, (Fuseable.ConditionalSubscriber<R>) outer)
-                : new SwitchOnFirstInnerSubscriber<>(this, outer);
+                ? new SwitchOnFirstConditionalControlSubscriber<>(this, (Fuseable.ConditionalSubscriber<R>) outer, cancelSourceOnComplete)
+                : new SwitchOnFirstControlSubscriber<>(this, outer, cancelSourceOnComplete);
             this.transformer = transformer;
         }
 
         @Override
         @Nullable
         public Object scanUnsafe(Attr key) {
-            if (key == Attr.CANCELLED) return cancelled;
-            if (key == Attr.TERMINATED) return done || cancelled;
+            final boolean isCancelled = this.inner == Operators.emptySubscriber();
+
+            if (key == Attr.CANCELLED) return isCancelled && !this.done;
+            if (key == Attr.TERMINATED) return this.done || isCancelled;
 
             return InnerOperator.super.scanUnsafe(key);
         }
 
         @Override
         public CoreSubscriber<? super R> actual() {
-            return outer;
-        }
-
-        @Override
-        public Context currentContext() {
-            CoreSubscriber<? super T> actual = inner;
-
-            if (actual != null) {
-                return actual.currentContext();
-            }
-
-            return outer.currentContext();
+            return this.outer;
         }
 
         @Override
         public void cancel() {
-            if (!cancelled) {
-                cancelled = true;
-                s.cancel();
+            if (INNER.getAndSet(this, Operators.emptySubscriber()) != Operators.emptySubscriber()) {
```

Comment:
Using empty subscriber instance (which I assume for internal usage only) in order to indicate canceled state and

Reply:
I guess to be extra sure to be shielded from `Operators.emptySubscriber()` ever returning a new instance, you can directly use `Operators.EMPTY_SUBSCRIBER` ? (fine for usage internal to the publisher package)